pyomo/contrib/community_detection/detection.py

```diff
@@ -0,0 +1,643 @@
+
+            # Find top_nodes (one of the two "groups" of nodes in a bipartite graph), which will be used to
+            # determine the graph layout
+            if model_graph.number_of_nodes() > 0 and nx.is_connected(model_graph):
+                # An index of 1 used because this tends to place constraint nodes on the left, which is
+                # consistent with the else case
+                top_nodes = nx.bipartite.sets(model_graph)[1]
+            else:
+                top_nodes = {node for node in model_graph.nodes() if node in constraint_variable_map}
+
+            if pos is None:  # The case where the user has not provided their own layout
+                pos = nx.bipartite_layout(model_graph, top_nodes)
+
+        else:  # This covers the case that type_of_community_map is 'constraint' or 'variable'
+
+            # Constraints are in the first list of the tuples in community map and variables are in the second list
+            position = 0 if type_of_graph == 'constraint' else 1
+            list_of_node_lists = list(i[position] for i in numbered_community_map.values())
+
+            # list_of_node_lists is (as it implies) a list of lists, so we will use the list comprehension
+            # below to flatten the list and get our one-dimensional node list
+            node_list = [node for sublist in list_of_node_lists for node in sublist]
+
+            # Now, we will find the first community that a node appears in and color the node based on
+            # that community (in numbered_community_map, certain nodes may appear in multiple communities,
+            # and we have chosen to give preference to the first community a node appears in)
+            color_list = []
+            for node in node_list:
+                not_found = True
+                for community_key in numbered_community_map:
+                    if not_found and node in numbered_community_map[community_key][position]:
+                        color_list.append(community_key)
+                        not_found = False
+
+            # Note - there is no strong reason to choose spring layout; it just creates relatively clean graphs
+            if pos is None:  # The case where the user has not provided their own layout
+                pos = nx.spring_layout(model_graph)
+
+        # Define color_map
+        color_map = cm.get_cmap('viridis', len(numbered_community_map))
+
+        # Create the figure and draw the graph
+        fig = plt.figure()
+        nx.draw_networkx_nodes(model_graph, pos, nodelist=node_list, node_size=40, cmap=color_map,
+                               node_color=color_list)
+        nx.draw_networkx_edges(model_graph, pos, alpha=0.5)
+
+        # Make the main title
+        graph_type = type_of_graph.capitalize()
+        community_map_type = self.type_of_community_map.capitalize()
+        main_graph_title = "%s graph - colored using %s community map" % (graph_type, community_map_type)
+
+        main_font_size = 14
```

Comment:
I don't think it is a good practice to fix font sizes here. Consider removing this and the one on line 457.